platform/lib/routers/search.js

```diff
@@ -0,0 +1,268 @@
+
+const REMOVE_HOST_PATTERN = /^https?:\/\/amp\.dev(?=\/)/;
+
+const PREVIEW_HOST = config.hosts.preview.base;
+const PLAYGROUND_URL = config.hosts.playground.base;
+
+// use the twitter title if available since it does not contain the site name
+const TITLE_META_TAG = 'twitter:title';
+const DESCRIPTION_META_TAG = 'twitter:description';
+
+// eslint-disable-next-line new-cap
+const examples = express.Router();
+
+const BUILD_IN_COMPONENTS = [
+  'amp-layout',
+  'amp-img',
+  'amp-pixel',
+];
+
+const IMPORTANT_INCLUDED_ELEMENTS = [
+  'amp-state',
+  'amp-list-load-more',
+  'amp-story-page',
+  'amp-story-grid-layer',
+  'amp-story-bookend',
+];
+
+const COMPONENT_VERSIONS_PATH = path.join(project.paths.GROW_POD,
+    '/extensions/amp-component-versions.json');
+
+function getAutosuggestComponents() {
+  let result = [];
+  const components = require(COMPONENT_VERSIONS_PATH);
+  for (const component in components) {
+    if (components.hasOwnProperty(component)) {
+      result.push(component);
+    }
+  }
+  result = result.concat(BUILD_IN_COMPONENTS);
+  result = result.concat(IMPORTANT_INCLUDED_ELEMENTS);
+  return result.sort();
+}
+
+const AUTOSUGGEST_COMPONENTS = getAutosuggestComponents();
+
+examples.get('/search/autosuggest', handleAutosuggestRequest);
+examples.get('/search/do', handleSearchRequest);
+
+function handleAutosuggestRequest(request, response) {
+  const items = ({
+    items: AUTOSUGGEST_COMPONENTS,
+  });
+  response.json(items);
+}
+
+function hasExample(component) {
+  const filePath = path.join(EXAMPLE_FOLDER_PATH, component + '.html');
+  return fs.existsSync(filePath);
+}
+
+function getCseItemMetaTagValue(item, metaTag) {
+  // since pagemap has always key:array the metatags dictionary is always the first element in the array
+  if (item.pagemap && item.pagemap.metatags && item.pagemap.metatags.length > 0
+      && item.pagemap.metatags[0][metaTag]) {
+    return item.pagemap.metatags[0][metaTag];
+  }
+  return null;
+}
+
+
+function createPageObject(csePageItem) {
+  let title = getCseItemMetaTagValue(csePageItem, TITLE_META_TAG);
+  if (!title) {
+    title = csePageItem.title;
+  }
+
+  const page = {
+    title: title,
+    description: csePageItem.snippet,
+    url: csePageItem.link,
+  };
+
+  const hostMatch = page.url.match(REMOVE_HOST_PATTERN);
+  if (hostMatch) {
+    page.url = page.url.substring(hostMatch[0].length);
+  }
+
+  return page;
+}
+
+function enrichComponentPageObject(item, page, locale) {
+  const description = getCseItemMetaTagValue(item, DESCRIPTION_META_TAG);
+  if (description) {
+    page.description = description;
+  }
+
+  if (page.url) {
+    const componentName = page.url.match(COMPONENT_REFERENCE_DOC_PATTERN)[1];
+    if (hasExample(componentName)) {
+      page.exampleUrl = EXAMPLE_URL_PATH + componentName + '/';
+      page.playgroundUrl = PLAYGROUND_URL + '/?url='
+        + encodeURIComponent(PREVIEW_HOST + page.exampleUrl);
```

Comment:
we should extract this into a helper method.

Reply:
I am not quite sure which part you mean

Reply:
`PLAYGROUND_URL + '/?url='
        + encodeURIComponent(PREVIEW_HOST + page.exampleUrl)` where we construct the playground URL. I think this is used in a few places. Maybe move it into config.js.

Reply:
We have a samples.js utility now